tensorboard/backend/event_processing/data_provider_test.py

```diff
@@ -0,0 +1,162 @@
+    multiplexer = event_multiplexer.EventMultiplexer()
+    multiplexer.AddRunsFromDirectory(self.logdir)
+    multiplexer.Reload()
+    return data_provider.MultiplexerDataProvider(multiplexer)
+
+  def test_list_scalars_all(self):
+    provider = self.create_provider()
+    result = provider.list_scalars(
+        experiment_id="unused",
+        owner_plugin=scalar_metadata.PLUGIN_NAME,
+        run_tag_filter=None,
+    )
+    self.assertItemsEqual(result.keys(), ["polynomials", "waves"])
+    self.assertItemsEqual(result["polynomials"].keys(), ["square", "cube"])
+    self.assertItemsEqual(result["waves"].keys(), ["square", "sine"])
+    sample = result["polynomials"]["square"]
+    self.assertIsInstance(sample, base_provider.ScalarMetadata)
+    self.assertEqual(sample.max_step, 18)
+    # nothing to test for wall time, as it can't be mocked out
+    self.assertIsInstance(sample.summary_metadata, summary_pb2.SummaryMetadata)
+    self.assertEqual(
+        sample.summary_metadata.plugin_data.plugin_name,
+        scalar_metadata.PLUGIN_NAME,
+    )
+
+  def test_list_scalars_filters(self):
+    provider = self.create_provider()
+
+    result = provider.list_scalars(
+        experiment_id="unused",
+        owner_plugin=scalar_metadata.PLUGIN_NAME,
+        run_tag_filter=base_provider.RunTagFilter(tags=["square", "quartic"]),
+    )
+    self.assertItemsEqual(result.keys(), ["polynomials", "waves"])
+    self.assertItemsEqual(result["polynomials"].keys(), ["square"])
+    self.assertItemsEqual(result["waves"].keys(), ["square"])
+
+    result = provider.list_scalars(
+        experiment_id="unused",
+        owner_plugin=scalar_metadata.PLUGIN_NAME,
+        run_tag_filter=base_provider.RunTagFilter(runs=["waves", "hugs"]),
+    )
+    self.assertItemsEqual(result.keys(), ["waves"])
+    self.assertItemsEqual(result["waves"].keys(), ["sine", "square"])
+
+    result = provider.list_scalars(
+        experiment_id="unused",
+        owner_plugin=scalar_metadata.PLUGIN_NAME,
+        run_tag_filter=base_provider.RunTagFilter(["un"], ["likely"]),
+    )
+    self.assertEqual(result, {})
+
+  def test_read_scalars(self):
+    provider = self.create_provider()
+
+    run_tag_filter = base_provider.RunTagFilter(
+        runs=["waves", "polynomials"], tags=["sine", "square", "cube"]
+    )
+    step_filter = base_provider.StepFilter(lower_bound=2, upper_bound=-3)
+    # steps [0, 2, ..., 18] from 2 to -3 = [2, ..., 16]
+    result = provider.read_scalars(
+        experiment_id="unused",
+        owner_plugin=scalar_metadata.PLUGIN_NAME,
+        run_tag_filter=run_tag_filter,
+        step_filter=step_filter,
+        downsample_to=None,  # not yet implemented
+    )
+
+    self.assertItemsEqual(result.keys(), ["polynomials", "waves"])
+    self.assertItemsEqual(result["polynomials"].keys(), ["square", "cube"])
+    self.assertItemsEqual(result["waves"].keys(), ["square", "sine"])
+    sample_series = result["polynomials"]["square"]
```

Comment:
Perhaps we could load data from the underlying multiplexer and verify that this data matches the relevant subset?  Then we wouldn't have to rely only on the spot-checking (but wouldn't have to tediously hardcode the expected summary data output).

Reply:
Done.